Replace django conf url imports

gherceg opened the pull request:
## Product Description
<!-- For non-invisible changes, describe user-facing effects. -->

## Technical Summary
<!--
    Provide a link to the ticket or document which prompted this change,
    Describe the rationale and design decisions.
-->
Easiest to review by commit.

The first two commits are straightforward changes since django 4.0 removed support for the import path we are currently using. `patterns` was deprecated in [v1.8 of django](https://docs.djangoproject.com/en/5.0/releases/1.8/#django-conf-urls-patterns), but I guess wasn't removed until at least Django 4.0. The docs for that do a good job of explaining the evolution of `patterns`, why it eventually became useless, and why the change made here is safe.

Notably, the flake8 changes in 3900f3558313cc88850bd0f2e4a2e5c9936e404d are more significant than usual. Specifically, the change to `download_index_files` and `source_files`. It looks like we no longer needed to serialize the app to json in `source_files`, because we only did this previously to append these results to the `files` variable (201e380c1f6c). Additionally, the returned list of files is already sorted in `download_index_files`, so the sort in `source_files` was redundant.

## Feature Flag
<!-- If this is specific to a feature flag, which one? -->

## Safety Assurance

### Safety story
<!--
Describe how you became confident in this change, such as
local testing, why the change is inherently safe, and/or plans to limit the blast radius of a defect.

In particular consider how existing data may be impacted by this change.
-->

### Automated test coverage

<!-- Identify the related test coverage and the tests it would catch -->

### QA Plan

<!--
- Describe QA plan that along with automated test coverages proves this PR is regression free
- Link to QA Ticket
-->
N/A

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

gherceg committed:
Replace django.conf.urls with django.urls

gherceg committed:
Remove patterns in favor of simple list

This was deprecated in 1.8 and supposed to removed in 2.0, but here we
are.

gherceg committed:
Fix flake8 warnings

gherceg commented on corehq/apps/sms/urls.py:
```diff
@@ -36,11 +36,13 @@
-    url(r'^add_gateway/(?P<hq_api_id>[\w-]+)/$',
-        AddDomainGatewayView.as_view(), name=AddDomainGatewayView.urlname
+    url(
+        r"^add_gateway/(?P<hq_api_id>[\w-]+)/$", AddDomainGatewayView.as_view(), name=AddDomainGatewayView.urlname
```
apologies for the change in quotation marks. Black is out of control.

mkangia commented on corehq/ex-submodules/casexml/apps/case/urls.py:
```diff
@@ -1,8 +1,8 @@
 from .views import reference_case_attachment_view
 
-urlpatterns = patterns('',
+urlpatterns = [
```
just curious, what is the impact of this change? I am unable to understand what the empty string did before.

gherceg replied:
Good question and sorry for not making this clear in the commit/PR description.

My understanding is that the first argument in `patterns` serves as a prefix that can be used to reduce repetition when referencing views as strings in the list of urls. The [django docs](https://docs.djangoproject.com/en/5.0/releases/1.8/#django-conf-urls-patterns) might help solidify this as they have code examples, but here's the gist:
> the patterns() function takes a required initial prefix argument which is prepended to all views-as-strings in that set of urlpatterns

Since Django no longer encourages referencing views as strings, but instead referencing the class/functions directly, the prefix no longer serves a purpose and is just an empty string when using `patterns`. This is what led to deprecating `patterns` in the first place, and ultimately its removal.

mkangia replied:
Thank you.

millerdev commented on corehq/apps/app_manager/views/download.py:
```diff
@@ -408,21 +407,24 @@ def validate_form_for_build(request, domain, app_id, form_unique_id, ajax=True):
 
 def download_index_files(app, build_profile_id=None):
+    def needed_for_ccz(path):
+        if profiles:
```
Looks like `profiles` is undefined if `build_profile_id is not None`, which seems like it would result in a name error. Does that sound right?

gherceg replied:
It sure does! I should have caught this. I suppose I could initialize it to None where `prefix` is set to a default value.

gherceg replied:
48a0b4a95e4bf1a3764ccb48678874f2bbece0d9

gherceg committed:
Ensure profiles is defined

Also move definition of needed_for_ccz into conditional to reflect that
it is only used within that path
